### PR TITLE
[DPE-2320] fix monitor secret id parsing issue

### PR DESCRIPTION
**Pull request by delgod**
Sometimes secret id is represented in different ways https://bugs.launchpad.net/juju/+bug/2028402.
this PR workaround the juju issue.

### CODE REVIEW

**Comment by delgod** [comment]
tests passed on juju3 here - https://github.com/canonical/mongodb-k8s-operator/actions/runs/5631150298?pr=181

File: src/charm.py [hunk 1]
```diff
@@ -533,21 +534,26 @@ def _on_set_password(self, event: ActionEvent) -> None:
...
+        elif event_secret_id == unit_secret_id:
             scope = UNIT_SCOPE
         else:
-            logging.debug(
-                "Secret {event._id}:{event.secret.id} changed, but it's irrelevant for us"
-            )
+            logging.debug("Secret %s changed, but it's unknown", event_secret_id)
             return
-        logging.debug(f"Secret {event._id} for scope {scope} changed, refreshing")
+        logging.debug("Secret %s for scope %s changed, refreshing", event_secret_id, scope)
```
Comment (dratushnyy) [inline src/charm.py:556]: just curious - why change formatting for the string?

File: tests/integration/metrics_tests/test_metrics.py [hunk 2]
```diff
@@ -76,7 +76,7 @@ async def test_endpoints_new_password(ops_test: OpsTest):
     action = await leader_unit.run_action("set-password", **{"username": "monitor"})
     action = await action.wait()
     # wait for non-leader units to receive relation changed event.
-    time.sleep(3)
+    time.sleep(10)
```
Comment (dratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:79]: maybe we can move it to constant or re-use leader re-election timeout constant?

File: src/charm.py [hunk 3]
```diff
@@ -533,21 +534,26 @@ def _on_set_password(self, event: ActionEvent) -> None:
         )
 
     def _on_secret_remove(self, event):
-        logging.debug(f"Secret {event._id} seems to have no observers, could be removed")
+        logging.debug(f"Secret {event.secret.id} seems to have no observers, could be removed")
 
     def _on_secret_changed(self, event):
-        secret = event.secret
+        regex = re.compile(".*[^/][/:]")
```
Comment (dratushnyy) [inline src/charm.py:540]: I think it is a good idea to add a pattern what this re matches in the comment

Comment (dmitry-ratushnyy) [inline src/charm.py:540]: I think it is a good idea to add a pattern what this re matches in the comment

File: src/charm.py (hunk 1 above)
Comment (dmitry-ratushnyy) [inline src/charm.py:556]: just curious - why change formatting for the string?

File: tests/integration/metrics_tests/test_metrics.py (hunk 2 above)
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:79]: maybe we can move it to constant or re-use leader re-election timeout constant?

File: src/charm.py (hunk 3 above)
Comment (delgod) [reply src/charm.py:546]: fixed

**Review by dmitry-ratushnyy — DISMISSED** [review]

File: src/charm.py (hunk 3 above)
Comment (dmitry-ratushnyy) [reply src/charm.py:546]: nits: I mean to write what this re will match, since not everyone good with getting the string out of pattern 😄

Comment (delgod) [reply src/charm.py:549]: added

**Review by dmitry-ratushnyy — DISMISSED** [review]

**Review by shayancanonical — APPROVED** [review]
Great regex!

File: src/charm.py [hunk 4]
```diff
@@ -533,23 +534,39 @@ def _on_set_password(self, event: ActionEvent) -> None:
...
+        # secret://9663a790-7828-4186-8b21-2624c58b6cfe/citb87nubg2s766pab40
+        # secret:citb87nubg2s766pab40
+        regex = re.compile(".*[^/][/:]")
+        event_secret_id = regex.sub("", event.secret.id)
+        app_secret_id = regex.sub(
+            "", self.app_peer_data.get(Config.Secrets.SECRET_INTERNAL_LABEL, "")
+        )
+        unit_secret_id = regex.sub(
+            "", self.unit_peer_data.get(Config.Secrets.SECRET_INTERNAL_LABEL, "")
+        )
```
Comment (juditnovak) [inline src/charm.py:556]: Don't we wanna have this logi hidden in a separate utility function or alike?

**Review by juditnovak — CHANGES_REQUESTED** [review]
I'm wondering if we should have a test (both for unit and app) which would have failed w/o the fix, and which is checking that the functionality of this fix is preserved in the future, no? It may be a bit complex as an integration test (due to eventual consistency), but should be well feasable as a unittest.

File: src/charm.py (hunk 4 above)
Comment (juditnovak) [reply src/charm.py]: Done.

File: src/charm.py [hunk 5]
```diff
@@ -533,23 +534,39 @@ def _on_set_password(self, event: ActionEvent) -> None:
         )
 
     def _on_secret_remove(self, event):
-        logging.debug(f"Secret {event._id} seems to have no observers, could be removed")
+        logging.debug(f"Secret {event.secret.id} seems to have no observers, could be removed")
 
     def _on_secret_changed(self, event):
-        secret = event.secret
+        """Handles passwords changes event.
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply src/charm.py]: Done.

**Review by juditnovak — APPROVED** [review]

File: src/charm.py (hunk 4 above)
Comment (delgod) [reply src/charm.py]: yes, please

File: src/charm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (delgod) [reply src/charm.py]: yes 🤦 it also should restart mongodb if tls secret changed 😞

**Comment by juditnovak** [comment]
Juju 3 pipeline: https://github.com/canonical/mongodb-k8s-operator/pull/183

**Comment by juditnovak** [comment]
Juju 3.2.0.1 tests passed here after the refactor: https://github.com/canonical/mongodb-k8s-operator/actions/runs/5693032447

**Review by dmitry-ratushnyy — APPROVED** [review]

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**